sdk/storage/azblob/internal/exported/structured_message_response_reader.go

```diff
@@ -0,0 +1,191 @@
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/internal/structuredmsg"
+)
+
+// StructuredMessageResponseReader wraps a response body and validates structured messages
+// if the x-ms-structured-body header is present. It strips the structured message framing
+// and returns only the validated data to the caller.
+type StructuredMessageResponseReader struct {
+	reader              *structuredmsg.StructuredMessageReader
+	underlyingBody      io.ReadCloser
+	currentSegmentData  []byte
+	currentSegmentIndex int
+	headerRead          bool
+	closed              bool
+}
+
+// NewStructuredMessageResponseReader creates a new reader that validates structured messages
+// if the structuredBodyType header indicates a structured message format.
+// If structuredBodyType is nil or empty, it returns the original body unchanged.
+func NewStructuredMessageResponseReader(body io.ReadCloser, structuredBodyType *string) (io.ReadCloser, error) {
+	if structuredBodyType == nil || *structuredBodyType == "" {
+		// No structured message, return body as-is
+		return body, nil
+	}
+
+	// Check if it's a structured message format we support
+	// Format: "XSM/1.0;CRC64" or "XSM/1.0;properties=crc64"
+	bodyType := strings.ToUpper(*structuredBodyType)
+	if !strings.HasPrefix(bodyType, "XSM/1.0") {
+		return body, nil
+	}
+
+	// Create structured message reader
+	reader := structuredmsg.NewStructuredMessageReader(body)
+
+	return &StructuredMessageResponseReader{
+		reader:             reader,
+		underlyingBody:     body,
+		currentSegmentData: nil,
+		headerRead:         false,
+		closed:             false,
+	}, nil
+}
+
+// Read implements io.Reader. It reads data from structured message segments,
+// validates CRC64, and returns only the data (stripping framing).
+func (r *StructuredMessageResponseReader) Read(p []byte) (int, error) {
+	if r.closed {
+		return 0, io.EOF
+	}
+
+	// Read header on first read
+	if !r.headerRead {
+		_, err := r.reader.ReadHeader()
+		if err != nil {
+			if err == io.EOF {
+				return 0, io.EOF
+			}
+			// Map structured message errors to blob errors
+			if errors.Is(err, structuredmsg.ErrInvalidVersion) {
+				return 0, fmt.Errorf("invalid structured message version: %w", err)
+			}
+			if errors.Is(err, structuredmsg.ErrUnexpectedEOF) {
+				return 0, fmt.Errorf("unexpected end of structured message: %w", err)
+			}
+			return 0, err
+		}
+		r.headerRead = true
+	}
+
+	// Read data from current segment, or get next segment
+	totalRead := 0
+	for totalRead < len(p) {
+		// If we've exhausted current segment, read next one
+		if r.currentSegmentData == nil || r.currentSegmentIndex >= len(r.currentSegmentData) {
+			segmentData, err := r.reader.ReadSegment()
+			if err != nil {
+				if err == io.EOF {
+					// No more segments, validate trailer
+					_, trailerErr := r.reader.ReadTrailer()
+					if trailerErr != nil {
+						if trailerErr == structuredmsg.ErrCRC64Mismatch {
```

Comment:
I think this (and other places) can be simplified.  If `ReadTrailer()` were to return the complete error message (e.g. "CRC64 checksum mismatch in structured message trailer"), there would be no need to export this error type and perform these comparisons.  You can simply return the error from `ReadTrailer()`.